src/object/obj_rpc.h

```diff
@@ -127,6 +127,7 @@ struct daos_obj_shard_tgt {
 	((uint32_t)		(orw_map_ver)		CRT_VAR) \
 	((uint32_t)		(orw_nr)		CRT_VAR) \
 	((daos_key_t)		(orw_dkey)		CRT_VAR) \
+	((uint32_t)		(orw_start_shard)	CRT_VAR) \
```

Comment:
Please make sure members are 64-bit aligned, you can just move orw_flags after this line
```
uint32_t orw_map_ver;
uint32_t orw_nr;
uint32_t orw_start_shard;
uint32_t orw_flags;
daos_key_t  orw_dkey;

Reply:
why need to add orw_start_shard?
for current full stripe update, shard_shard is always 0, no? it it is true, we probably need not worry about it now and leave it to later.

Reply:
Start shard is not always zero. This is why I added the RPC parameter.  We talked about this in a meeting, and I recall that everyone (you, WangDi, Liang, Johann) all agreed that I should add this.

Without it, the targets cannot determine where they are in the stripe.

Reply:
Sorry I am still not very clear why we need orw_start_shard.
I think each target need to know the offset within the full stripe, for example for e_k == 8, then each target need to know it is the 0, or 1, 2... 7 target in the full stripe.
When server process the update request, by check object layout it can know its shard_number, and (shard_number / e_k) is the offset it needs to know. Why still need that orw_start_shard?

Reply:
This was also discussed in the meeting held on 3/7/2019 at 7:00am MST.  The entire team (you, WangDi, Liang, Johann) approved this approach and agreed that the start shard should be sent in the RPC.  There was not dissent or disagreement from anyone in that meeting.

I had looked at the object layout prior to the 3/7/19 meeting and I determined that the function calls available to determine the object layout were all designed to be run within the client, and hence this approach would require dc_ (client based) representations of the object that are not used with the server. Because of this difficulty, everyone in the meeting agreed that just sending the start shard in the RPC would be preferred approach. 

To summarize, the shard number sent in the RPC is not indexed based on the redundancy group and cannot be mapped to 0...k-1 without knowing the index of the starting shard in the redundancy group.  Although it is probably possible to determine the index of the starting shard in the redundancy group on the server by somehow looking this value up in the object layout, it was decided that since the client already has this information, that it should just be sent in the RPC.

Reply:
Thanks for explanation. I do remember we ever discussed that, but in my memory the approach was that when leader server dispatch the IO req to different target, the leader will tell each target the shard index of that target. 
So this orw_start_shard is for the client tell the leader the shart_shart, so when leader dispatch it can calculate the shard offset/index and make it known for the non-leader targets? is it what you plan to do? 
just that did not see other related code in this patch, so just to confirm ...

Reply:
The meeting was a long time ago, so it's understandable that you don't remember the discussion.

The leader does tell each target it's shard_index, but this is not sufficient (since it's not zero-relative) to know position in the stripe.  There is no EC-specific handling planned for the leader (WangDi insists on this). Instead, the start_shard is forwarded in the RPC call to the target (the leader just forwards the orw received from the client and only changes a few parameters (e.g., shard_index).  The targets will then compute shard_index - start_shard to determine their position in the stripe.